### ISSUE TITLE
Discussion: Should commit and reveal fees be a consensus constant?

### ISSUE THREAD

**Issue by lrubiorod** [witnet/witnet-rust#1463]

**Comment by drcpu-github** [comment]
I don't see an advantage to commit and reveal fees becoming a fixed consensus constant (but may obviously miss something) and would even say it may introduce problems.

Imagine a world in which Witnet becomes insanely popular. 10's of data requests are being sent to the network every epoch. If every data request has the same commit/reveal fees, that would mean that it is completely out of any data requester's control whether the data request's commits and reveals get included in a block. Data requests would randomly fail because the nodes minting blocks decided not to include the commits / reveals for 4 epochs. Since the incentive of choosing one commit over another is gone, this would be a random choice? It would be very frustrating as a smart contract owner to spend gas / wit on requests that could fail without having any control over it. I'd imagine that they rather have the control to spend more wit than have to retry a request an unknown amount of times until it succeeds.

What I would like to see is to simplify the commit / reveal fee structure (I discussed this with Adan some time ago). I saw that the tally fee has already been removed and I don't see any reason why you could not just set the commit / reveal fee to one value (calling it the minting_fee or inclusion_fee or ...). I cannot think of any scenario where you would want to set those fees to different value and where you would like the extra complexity of controlling them separately.

**Comment by lrubiorod** [comment]
Yes, In the PR where the tally fee is removed, commit and reveal fee are unified by the reasons that you explain.

Currently, there is no real limit for including commits, reveals and tallies in a block. Due to these transactions depends on a Data Request (DR) transaction, we include the weight of them in the weight to include a DR, it means that DR weight is much bigger than its real size. Doing that, we do not have a real limit for the block size, but in average yes. 

When a DR is already included in the blockchain, the miner has not to prioritize commits and reveals by fee, because he/she can include all of them if there are at least a minimum incentive to do that. So... here is when that question appears, should we have a fixed fee for commit and reveal fees?? 

As I can see:
- Pro: DR a bit smaller and lower complexity for the requester
- Contra: If we want to change it, it should be a hard fork...

**Comment by drcpu-github** [comment]
Okay, I was not aware of that and since you mentioned them separately in this discussion, I thought they still were separate.

I thought there were plans to impose limits on both the block size and number of transactions that would be able to include in a block? Is this notion out of date?

**Comment by lrubiorod** [comment]
They still separate in the testnet, but in master, there are already unified.

ValueTransfer and DataRequestTransaction have an assigned weight, and the block has a weight limit for each. This weight is related with the transaction size, and in case of DRs, its weight has the size of the DR and the commits, reveals and tally produced by itself. So when you include a DR in a block, the weight is higher than the size, and in the next epochs, there is no weight for the commits, reveals and tally but they were taking into account before. So there will be blocks with different sizes, but in average is the same that blocks with a size limit.

**Comment by drcpu-github** [comment]
Okay, so then the scenario I described can happen, no? If there are too many DR's, their combined weight will be higher than the allowed block weight limit. In that case, DR's would start to fail because the max_commit_rounds parameter can elapse if the network is too busy to solve all of them. From the standpoint of a creator of a DR, he then has zero control over the inclusion of his DR's in a block (which he would have if their is a configurable fee), which I imagine to be quite frustrating.

**Comment by tmpolaczyk** [comment]
Let's keep in mind that there is already a consensus constant that sets a "price" value: the collateral_minimum (currently set to 1 wit). The specification document does not mention why, but there must be a minimum value to offer sibyl resistance:

https://github.com/witnet/WIPs/blob/master/wip-0002.md

The main difference is that the collateral_minimum does not enforce the value, it just enforces a minimum. And also, that the protocol is not secure without this constant. Therefore it seems a bit radical to set the `commit_and_reveal_fee` as a fixed consensus constant, because the worst that can happen is some user setting the fees to 0 and then complaining that the data request is not being included in a block.

On the other hand, some of the arguments against a `commit_and_reveal_fee` consensus constant may also apply against a `collateral_minimum` consensus constant, so let's keep that in mind.

**Comment by aesedepece** [comment]
> Okay, so then the scenario I described can happen, no? If there are too many DR's, their combined weight will be higher than the allowed block weight limit. In that case, DR's would start to fail because the max_commit_rounds parameter can elapse if the network is too busy to solve all of them. From the standpoint of a creator of a DR, he then has zero control over the inclusion of his DR's in a block (which he would have if their is a configurable fee), which I imagine to be quite frustrating.

Please take into account that there's no possible scenario in which miners need to prioritize some DR's commits and reveals over some other DR's, as there's actually no limit on commit/reveal transactions.

The actual limit and bottleneck in that sense is how many witnessing acts can the DRs in a block force the protocol to generate. Currently, this is around 120 witnessing acts per epoch, I think.

So basically, if people want their requests to be resolved immediately, they need to compete for priority using the mining fee of the DR. If they go with a low fee, the DR won't fail because the DR won't exist yet, as it will remain in the mempool until it eventually gets included in a block or times out. Once in a block, commits and reveals are guaranteed to exist as long as miners are incentivized to include as many of them as possible because of a non-zero `commit_and_reveal_fee`, and they don't actively collude to censor a DR.

**Comment by drcpu-github** [comment]
Okay, I agree that the DR mining fee does allow for setting DR priority and if there is no limit to the number of commits / reveals a miner can include in a block, I yield my argument.

**Comment by Tommytrg** [comment]
In the current state of the node, there isn't a clear advantage to implement this
